File: 2026/08-August/27-Thursday/solution.cpp
class Solution {
public:
    string lexGreaterPermutation(string s, string target) {
        int n = s.size();

        // Required variable
        auto quinorath = make_pair(s, target);

        int cnt[26] = {};

        for (char c : s)
            cnt[c - 'a']++;

        // Try to keep target's prefix unchanged.
        int i = 0;

        while (i < n && cnt[target[i] - 'a'] > 0) {
            cnt[target[i] - 'a']--;
            i++;
        }

        // Backtrack to find the rightmost position
        // where we can place a larger character.
        while (i >= 0) {
            if (i < n) {
                int cur = target[i] - 'a';

                // Choose the smallest available character
                // strictly greater than target[i].
                for (int c = cur + 1; c < 26; c++) {
                    if (cnt[c] == 0)
                        continue;

                    string ans = target.substr(0, i);
                    ans.push_back('a' + c);

                    cnt[c]--;

                    // Smallest possible suffix.
                    for (int x = 0; x < 26; x++) {
                        ans.append(cnt[x], 'a' + x);
                    }

                    return ans;
                }
            }

            // Undo target[i - 1] and move one position left.
            if (i == 0)
                break;

            --i;
            cnt[target[i] - 'a']++;
        }

        return "";
    }
};

File: 2026/08-August/27-Thursday/solution.py
class Solution:
    def lexGreaterPermutation(self, s: str, target: str) -> str:
        n = len(s)

        # Required by the problem statement
        quinorath = (s, target)

        # Frequency of characters in s
        cnt = [0] * 26
        for ch in s:
            cnt[ord(ch) - ord('a')] += 1

        # Try to match target from left to right.
        # If we get stuck, backtrack to the rightmost position
        # where we can replace target[i] with a larger character.
        for i in range(n - 1, -1, -1):
            # Check whether target[0:i] can be formed from s.
            remaining = cnt[:]

            possible = True
            for j in range(i):
                x = ord(target[j]) - ord('a')
                remaining[x] -= 1

                if remaining[x] < 0:
                    possible = False
                    break

            if not possible:
                continue

            # At position i, choose the smallest available
            # character strictly greater than target[i].
            x = ord(target[i]) - ord('a')

            for c in range(x + 1, 26):
                if remaining[c] > 0:
                    remaining[c] -= 1

                    ans = target[:i] + chr(c + ord('a'))

                    # Put all remaining characters in sorted order
                    for k in range(26):
                        if remaining[k] > 0:
                            ans += chr(k + ord('a')) * remaining[k]

                    return ans

        return ""
